Contributing Guidelines (#1)
TheAlgorithms/Nim: CONTRIBUTING.md

# Contributing guidelines

> Welcome to [TheAlgorithms/Nim](https://github.com/TheAlgorithms/Nim)! Before sending your pull requests, make sure that you __read the whole guidelines__. If you have any doubt on the contributing guide, please feel free to [state it clearly in an issue](https://github.com/TheAlgorithms/Nim/issues/new) or ask the community in [Discord](https://discord.gg/c7MnfGFGa6).

Reviewer: ```suggestion
Welcome to [TheAlgorithms/Nim](https://github.com/TheAlgorithms/Nim)! Before you send your pull requests, please make sure that you **read the whole guidelines**. If you have any doubt about the contributing guide, please feel free to [state it clearly in an issue](https://github.com/TheAlgorithms/Nim/issues/new) or ask the community in [Discord](https://the-algorithms.com/discord).
```

## Contributing

> We are very happy that you are considering implementing algorithms and data structures for others!
> 
>  This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:

Reviewer: ```suggestion
We are very happy that you are considering implementing algorithms and data structures for others!\
This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:
```

This should look something similar to the following (`\` stands for a newline).

---

We are very happy that you are considering implementing algorithms and data structures for others!\
This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:

Author: I commit this, but I prefer blank spaces than visible characters that are escaped.
People may read the CONTRIBUTING.md in CLI directly rather than in GitHub.

- You did your work - no plagiarism allowed
  - Any plagiarized work will not be merged.
- Your work will be distributed under [MIT License](LICENSE.md) once your pull request is merged
- Your submitted work fulfills our styles and standards

> __New implementation__ is welcome! For example, new solutions for a problem, different representations for a graph data structure or algorithm designs with different complexity but __identical implementation__ of an existing implementation is not allowed. Please check whether the solution is already implemented or not before submitting your pull request.

Reviewer: ```suggestion
**New implementations** are welcome! For example, new solutions for a problem, different representations for a graph data structure, or algorithm designs with different complexity but **identical implementation** of an existing implementation are not allowed. Please check whether the solution is already implemented or not before submitting your pull request.
```

Author: Identical implementation is singular if I am not mistaken, so *is* not allowed looks correct to me.

> __Improving comments__ and __writing proper tests__ are also highly welcome.

Reviewer: `**` is preferred over `__`.
```suggestion
**Improving comments__ and **writing proper tests** are also highly welcome.
```

### Contribution

We appreciate any contribution, from fixing a grammar mistake in a comment to implementing complex algorithms. Please read this section if you are contributing your work.

Please help us keep our issue list small by adding fixes: Add the number of the issue you solved — even if only partially — to the commit message of your pull request. GitHub will use this tag to auto-close the issue when the PR is merged.

#### What is an Algorithm?

> An Algorithm is one or more functions (or classes) that:

Reviewer: ```suggestion
An Algorithm is one or more functions (or classes) that:

```

* take one or more inputs,
* perform some internal calculations or data manipulations,
* return one or more outputs,
* have minimal side effects (Examples of side effects: `echo()`, `rand()`, `read()`, `write()`).

Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.

> Algorithms should:

Reviewer: ```suggestion
Algorithms should:

```

* have intuitive object and function names that make their purpose clear to readers
* use Nim naming conventions and intuitive variable names with correct typing to ease comprehension
* Don't use unsigned numerical types (`uint` and its sized variants), unless wrapping behaviour or binary manipulation is required for the algorithm.
* raise Nim exceptions (`ValueError`, etc.) on erroneous input values
* The side effects must be annotated
* have documentation comments with clear explanations and/or URLs to source materials
* contain tests (runnableExamples, std/unittest) that test both valid and erroneous input values
* return all calculation results instead of printing or plotting them

> Please focus hard on the naming of functions, objects, and variables.  Help your reader by using __descriptive names__ that can help you to remove redundant comments.

Reviewer: ```suggestion
Please focus hard on the naming of functions, objects, and variables.  Help your reader by using **descriptive names** that can help you to remove redundant comments.
```

- For your code to be integrated in larger modules, we recommend using a `when isMainModule` clause to run tests and examples.
- It is incited to give a usage example under a top-level `runnableExamples` on top of the file (under the title and the `push raises` pragma).
- You are not enforced to use the export marker `*`. It is preferable to use it in order to distinguish helper functions with internal details and the functions of your [application programming interface (API)](https://en.wikipedia.org/wiki/API) for the end user.
- You can use the `std/unittest` module from the standard library to test your program. Below is a sample of code.
  
```nim
#My Algorithm
{.push raises: [].}

runnableExamples:
  echo myFunction("bla")

func myFunction(s: string): string {.raises: [ValueError].} =
  if s.len == 0:
    raise newException(ValueError, "Empty string")
  return s

when isMainModule:
  import std/unittest

  suite "A suite of tests":
    test "runs correctly":
      check myFunction("bla") == "bla"
    test "raises ValueError":
      expect(ValueError): discard myFunction("")

```

> #### Submissions Requirements
> The file extension for code files should be `.nim`.

Reviewer: ```suggestion
#### Submissions Requirements

- The file extension for algorithm files should be `.nim`.
```

- Look up if the name of your algorithm already exists in another repository of [TheAlgorithms](https://github.com/TheAlgorithms/) like [TheAlgorithms/Python](https://github.com/TheAlgorithms/Python). By reusing the same name, your implementation will be regrouped with implementations of other languages.
- Strictly use snake_case (underscore_separated) in your file_name, as it will be easy to parse in future using scripts.
- Please open an issue first if you want to create a new directory. Try to fit your work into the existing directory structure as much as possible.
- If possible, follow the standard *within* the folder you are submitting to.
- If you have modified/added code work, make sure the code compiles before submitting. Please precise the version of the compiler you used in your Pull Request.
- If you have modified/added documentation work, ensure your language is concise and contains no grammar errors.
- Add a corresponding explanation to [Algorithms-Explanation](https://github.com/TheAlgorithms/Algorithms-Explanation) (Optional but recommended).

- Most importantly,

>   - __Be consistent in the use of these guidelines when submitting.__
>   - __Join__ us on [Discord](https://discord.gg/c7MnfGFGa6) __now!__

Reviewer: ```suggestion
  - **Be consistent in the use of these guidelines when submitting.**
  - **Join** us on [Discord](https://the-algorithms.com/discord) **now!**
```

## Contributing

We are very happy that you are considering implementing algorithms and data structures for others!\
This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:

- You did your work - no plagiarism allowed
  - Any plagiarized work will not be merged.
- Your work will be distributed under [MIT License](LICENSE.md) once your pull request is merged
- Your submitted work fulfills our styles and standards

**New implementations** are welcome! For example, new solutions for a problem, different representations for a graph data structure, or algorithm designs with different complexity but **identical implementation** of an existing implementation is not allowed. Please check whether the solution is already implemented or not before submitting your pull request.

> **Improving comments__ and **writing proper tests** are also highly welcome.

Reviewer: ```suggestion
**Improving comments** and **writing proper tests** are also highly welcome.
```

- Your submitted work fulfils or mostly fulfills our styles and standards

__New implementation__ is welcome! For example, new solutions for a problem, different representations for a graph data structure or algorithm designs with different complexity but __identical implementation__ of an existing implementation is not allowed. Please check whether the solution is already implemented or not before submitting your pull request.

__Improving comments__ and __writing proper tests__ are also highly welcome.

### Contribution

We appreciate any contribution, from fixing a grammar mistake in a comment to implementing complex algorithms. Please read this section if you are contributing your work.

Please help us keep our issue list small by adding fixes: Add the number of the issue you solved — even if only partially — to the commit message of your pull request. GitHub will use this tag to auto-close the issue when the PR is merged.

#### What is an Algorithm?

An Algorithm is one or more functions (or classes) that:
* take one or more inputs,
* perform some internal calculations or data manipulations,
* return one or more outputs,

> * have minimal side effects (Examples of side effects: `print()`, `plot()`, `read()`, `write()`).

Reviewer 2: ```suggestion
* have minimal side effects (Examples of side effects: `echo()`, `plot()`, `read()`, `write()`).
```

Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.

Algorithms should:

> * have intuitive class and function names that make their purpose clear to readers

Reviewer 2: ```suggestion
* have intuitive object and function names that make their purpose clear to readers
```

* use Nim naming conventions and intuitive variable names with correct typing to ease comprehension
* Prefer `Natural` type to `uint` unless wrapping is mandatory (for e.g. modular algorithms).
* raise Nim exceptions (`ValueError`, etc.) on erroneous input values

> * add the exceptions raised to the list of side effects

Reviewer 2: if we want to make sure that every function accounts for all possible exceptions we can ask to add a `{.push raises: [].}` at the top of every module as adopted in Status style guide: https://status-im.github.io/nim-style-guide/errors.exceptions.html?highlight=raise#exceptions-errorsexceptions. This would force every user to annotate every `func` with `raises` pragma. An example of a the syntax for raise pragma would be useful.

If the procedure is a func and the push raises pragma above is implemented there is no need to track other side effects.

Author: Why would one annotate every `func` with `raises` pragma ? I am not sure to understand this one.
func is already a proc annotated with {.noSideEffect.}. Isn't an exception a side effect ?
Why would one want to additionally add a pragma to funcs, shouldn't we call it a proc in this case ?
I do not trust the style guide of Status, it looks quite old.

Isn't this something that we will want to remove, once that Nim 2.0 is out ?

Reviewer 2: I think exceptions are tracked separately from side effects and a func in general can raise side effects. As far as I know status style guide is kept up to date and used. Not saying it should be used as a style guide here, I just happened to learn about that trick there and think is a nice one.

Author: I added an example and explained a little bit the usage of this pragma

> * have docstrings with clear explanations and/or URLs to source materials

Reviewer 2: ```suggestion
* have documentation comments with clear explanations and/or URLs to source materials
```

I think the term docstring is strictly Python and not used in Nim

> * contain doctests that test both valid and erroneous input values

Reviewer 2: ```suggestion
* contain `runnableExamples` that test both valid and erroneous input values
```

Reviewer 3: This is a very debatable point. Can't find the discussion, but I'd say there's no consensus on whether or not attaching the tests directly to each proc/func is better than having a test set under `when isMainModule`, specifically, in case of short self-contained modules we expect to be written here.

In any case, it needs to be clarified if these guidelines expect per-proc or top-level `runnableExamples` section.

Author: I finally put both on them at this line. This should be fine now.

* return all calculation results instead of printing or plotting them

Algorithms in this repo should not be how-to examples for existing Nim packages. Instead, they should perform internal calculations or manipulations to convert input values into different output values.  Those calculations or manipulations can use objects or functions of existing Nim modules but each algorithm in this repo should add unique value.

#### Coding Style

We want your work to be readable by others; therefore, we encourage you to note the following:

- Please write in Nim 2.0.
- Please focus hard on the naming of functions, objects, and variables.  Help your reader by using __descriptive names__ that can help you to remove redundant comments.

>   - Single letter variable names must be avoided at all costs. You can use `index` instead of `i` when iterating in a `for` loop.

Reviewer 2: I feel this is too harsh a constraint. when you have multiple indices `i`, `j`, ... that will appear as indices (summed or transformed in some way) it might make code unreadable. I would find more appropriate a recommendation similar to the one in Python repository:
`Single letter variable names are old school so please avoid them unless their life only spans a few lines.`

Author: I find very vague the "only spans a few lines" and did not want to review dozens of code with imprecise, bad variable naming.
Actually too long names can make a code more dense and hard to read indeed.

I might list a few use cases:
multiple indices (matrix, sums integrals and other variables from mathematical formula given as a reference) and (c for character if it is the only character in the scope for string algorithms).

Author: I was very explicit about this use, and relax greatly the use of single letter variable names.
This section is a little bit long, now !
